### ISSUE TITLE
Why ship generated files?

### ISSUE THREAD

**Issue by richsalz** [openssl/openssl#9966]
OpenSSL ships objects.txt and all the files that are built from it.  Is that still necessary?  It's hard to see the rationale for doing so, since so many other files are built at configure time now.
Ping @levitte since this might have out out-of-tree implications.

There are many PR's that have been bigger than they needed to be because of this, for example.

**Comment by beldmit** [comment]
And also many merge problems...

Not sure we can totally get rid of OIDs and NIDs,  though.

**Comment by richsalz** [comment]
I don't want to get rid of them, I just don't want to ship the files; run the perl scripts at the same time we build CA.pl, configdata, etc.

**Comment by levitte** [comment]
Just like other numeral files (`util/*.num`, `crypto/err/openssl.txt`, to name a few), they contain numbers that people expect to be constant over time. Among others, NIDs are shared between libcrypto and engines, and applications can use them too, so if those numbers would change from one openssl build to another, we would get complete chaos.

This is a very tangible reason why we try to avoid NIDs as much as possible in the new provider interface, so we don't fall into that same maintenance trap (and also, so provider vendors don't have to ask us to insert NIDs for their stuff, like we currently have to do with, for example, gost).

**Comment by richsalz** [comment]
That's a good comment, @levitte, but as long as OpenSSL controls objects.txt, the numbers won't change.  Your comment doesn't explain why the dependant files could not be built at configuration time.

**Comment by levitte** [comment]
The *NIDs* are public, constant and generated. They are *not* present in objects.txt. You are mixing them up with OIDs.

**Comment by beldmit** [comment]
Hmm. Sure, enforcing `make update` for freshly added objects does the trick.

**Comment by beldmit** [comment]
In theory, we could exclude autogenerated files from PRs to reduce the probability of merge conflicts, but it will make the final merge and travis tests more complicated.

**Comment by levitte** [comment]
Yeah, sure, it's perfectly possible to wait with the `make update` until release time...  that is, unless something new in the code uses the newly generated NIDs.

**Comment by richsalz** [comment]
The NIDs are public constant and generated.

OpenSSL controls objects.txt and the scripts that parse it and generate all the other things.

Why not do it at config time?

**Comment by richsalz** [comment]
> In theory, we could exclude autogenerated files from PRs to reduce the probability of merge conflicts, but it will make the final merge and travis tests more complicated.

No final merge.  The only file in the tree is objects.txt and the scripts that generate the files.

It gets even EASIER.

**Comment by levitte** [comment]
> Why not do it at config time?

Because existing NIDs need to remain constant across minor versions, let alone from one build to another.  It's the exact same problem as the numbers generated for `_R_` macros, do you want to generate those on the fly too?

**Comment by richsalz** [comment]
>Because existing NIDs need to remain constant across minor versions, let alone from one build to another.

Except that OpenSSL makes no checks to ensure that this happens.  If we wanted to do that, we should have a separate "state" file, just like the error macros.  (And, of course, the project has no tests to see if the error number state file *changed* anything, as opposed to *adding* anything.)  I might do a test across various versions to see if the NID's changed.

**Comment by levitte** [comment]
The "state" file you're looking for is `crypto/objects/obj_mac.num`

**Comment by richsalz** [comment]
Why is everything in "make generate" and "make update" done at make time, and not at Configuration time?  ```Configure``` is around 3200 lines of perl.  Do we not trust all the other perl scripts to run properly?

**Comment by beldmit** [comment]
I'm afraid the answer is the same. If software refers to NIDs, they should be constant. Do we have a way to make them constant without storing them in this or that form?

**Comment by richsalz** [comment]
Please see https://github.com/openssl/openssl/issues/9966#issuecomment-533818815

**Comment by richsalz** [comment]
If we ship the same state file and the same perl scripts, why do we not trust that everyone will get the same output?  And why do we trust it for Configure and various ```.in``` files, but not the other autogenerated files?

**Comment by beldmit** [comment]
Got it. If we ship the same state file, it works. It let us to regenerate of `obj_dat.h` and `obj_mac.h` in during Configure (or maybe, make) phase.

**Comment by levitte** [comment]
Aaaaah, that's what you were after. Yeah, good point

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev, barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
